extensions/azurecore/src/account-provider/auths/azureAuth.ts

```diff
@@ -134,13 +157,130 @@ export abstract class AzureAuth implements vscode.Disposable {
+		} catch (ex) {
+			if (ex instanceof AzureAuthError) {
+				void vscode.window.showErrorMessage(ex.message);
+				Logger.error(ex.originalMessageAndException);
+			} else {
+				Logger.error(ex);
+			}
+			Logger.error(ex);
+			account.isStale = true;
+			return account;
+		}
+	}
+
+	public async hydrateAccount(token: Token | AccessToken, tokenClaims: TokenClaims): Promise<AzureAccount> {
+		let account: azdata.Account;
+		if (this.authLibrary === 'ADAL') {
+			const tenants = await this.getTenants({ ...token });
+			account = this.createAccount(tokenClaims, token.key, tenants);
+		} else {
+			const tenants = await this.getTenantsMsal(token.token);
+			account = this.createAccount(tokenClaims, token.key, tenants);
+		}
 		return account;
 	}
 
-	protected abstract login(tenant: Tenant, resource: Resource): Promise<{ response: AuthenticationResult, authComplete: Deferred<void, Error> }>;
+	public async getAccountSecurityToken(account: AzureAccount, tenantId: string, azureResource: azdata.AzureResource): Promise<Token | undefined> {
+		if (account.isStale === true) {
+			Logger.error('Account was stale. No tokens being fetched.');
+			return undefined;
+		}
+
+		const resource = this.resources.find(s => s.azureResourceId === azureResource);
+		if (!resource) {
+			Logger.error('Invalid resource, not fetching', azureResource);
+
+			return undefined;
+		}
+
+		const tenant = account.properties.tenants.find(t => t.id === tenantId);
+
+		if (!tenant) {
+			throw new AzureAuthError(localize('azure.tenantNotFound', "Specified tenant with ID '{0}' not found.", tenantId), `Tenant ${tenantId} not found.`, undefined);
+		}
+
+		const cachedTokens = await this.getSavedToken(tenant, resource, account.key);
+
+		// Let's check to see if we can just use the cached tokens to return to the user
+		if (cachedTokens?.accessToken) {
+			let expiry = Number(cachedTokens.expiresOn);
+			if (Number.isNaN(expiry)) {
+				Logger.error('Expiration time was not defined. This is expected on first launch');
+				expiry = 0;
+			}
+			const currentTime = new Date().getTime() / 1000;
+
+			let accessToken = cachedTokens.accessToken;
+			let expiresOn = Number(cachedTokens.expiresOn);
+			const remainingTime = expiry - currentTime;
+			const maxTolerance = 2 * 60; // two minutes
+
+			if (remainingTime < maxTolerance) {
+				const result = await this.refreshToken(tenant, resource, cachedTokens.refreshToken);
+				if (result) {
+					accessToken = result.accessToken;
+					expiresOn = Number(result.expiresOn);
+				}
+			}
+			// Let's just return here.
+			if (accessToken) {
+				return {
+					...accessToken,
+					expiresOn: expiresOn,
+					tokenType: 'Bearer'
+				};
+			}
+		}
+
+		// User didn't have any cached tokens, or the cached tokens weren't useful.
+		// For most users we can use the refresh token from the general microsoft resource to an access token of basically any type of resource we want.
```

Comment:
I think the comments might have been misplaced, can you please review them if they apply to the respective code blocks?

Reply:
They aren't misplaced, this code was there previously as seen here: https://github.com/microsoft/azuredatastudio/blob/5f8cefc45c46c66625cfae5b83886c85ce6c172f/extensions/azurecore/src/account-provider/auths/azureAuth.ts#L228 I was just moving it around